### DISCUSSION TITLE
Clarification needed regarding field dumps and lumped ports

### DISCUSSION THREAD

**Discussion by ignitions2** [thliebig/openEMS-Project#70] [Q&A Usage]
Hi,
I am struggling with three questions:

1. I am studying the electric field, the magnetic field and the electric current density inside and outside two materials defined by boxes. The physical quantities are to be simulated in time domain. The excitation is done using lumped port. During the simulation the above mentionned fields are saved in particular planes defined by field dumps (DumpType=0, 5 and 2). The simulation looks okay, terminating without error, I was able to load and visualize the fields against time in paraview, however, the interpretation of the magnitude of the fields is not clear. Unfortunately no units are indicated on the colorbars in paraview. Do you know what the units of the simulated dump fields are and how to add them to colorbar in paraview? It would be good to understand this basically for all DumpTypes. My code can be seen below.

2. The second question is regarding lumped ports. As I mentionned above, for the excitation I am using a lumped port. However, it is not clear what the physical meaning of the following things:
- direction of the port (since the port is defined by two points, that would define a direction anyway, why is adding 'direction' separately needed?)
- amplitude of the port (When defining the direction, one can define the amplitude as well. But it is unclear, what the meaning of the amplitude of a port is, and what the physical unit of this 'amplitude' is.) 

3. And finally, can you please tell me (e.g. through some example) how the voltage probe and current probe is defined and used in case of lumped ports? How can one get the incoming and reflected current?

close all
clear all
clc

unit = 1e-6;

FDTD = InitFDTD('NrTS',5e5, 'EndCriteria',0, 'OverSampling',50);
FDTD = SetSinusExcite(FDTD,10e9);
FDTD = SetBoundaryCond(FDTD,{'MUR' 'MUR' 'MUR' 'MUR' 'MUR' 'MUR'});

CSX = InitCSX();

mesh.x = [-10:0.5:-3 -2.8:0.2:2.8 3:0.5:10];
mesh.y = [-10:0.5:-2 -1.8:0.2:1.8 2:0.5:10];
mesh.z = -10:110;
CSX = DefineRectGrid(CSX, unit, mesh);

CSX = AddMaterial( CSX, 'stripline1' );
CSX = SetMaterialProperty(CSX, 'stripline1','Epsilon',1,'Mue',0.999991,'Kappa',59594756);
start = [-2.6, -1, 0];
stop  = [-0.6, 1, 100];
CSX = AddBox( CSX, 'stripline1', 999, start, stop );

CSX = AddMaterial( CSX, 'stripline2' );
CSX = SetMaterialProperty(CSX, 'stripline2','Epsilon',1,'Mue',0.999991,'Kappa',59594756);
start = [0.6, -1, 0];
stop  = [2.6, 1, 100];
CSX = AddBox( CSX, 'stripline2', 999, start, stop );

start = [-1.6 0 0];
stop  = [1.6 0 0];
[CSX, port] = AddLumpedPort(CSX, 5 ,1 , 50, start, stop, [1e4 0 0], true);

%Checking E-field in x-z cross section
CSX = AddDump(CSX,'Et_x_z');
CSX = AddBox(CSX,'Et_x_z',0,[-10 0 -10],[10 0 110]);

%Checking E-field in x-y cross section
CSX = AddDump(CSX,'Et_x_y');
CSX = AddBox(CSX,'Et_x_y',0,[-10 -10 30],[10 10 30]);

%Checking B-field in x-y cross section
CSX = AddDump(CSX,'Bt_x_y','DumpType',5);
CSX = AddBox(CSX,'Bt_x_y',0,[-10 -10 30],[10 10 30]);

%Checking current in x-y cross section
CSX = AddDump(CSX,'Jt_x_y','DumpType',2);
CSX = AddBox(CSX,'Jt_x_y',0,[-10 -10 30],[10 10 30]);

mkdir('tmp');
WriteOpenEMS('tmp/tmp.xml',FDTD,CSX);

CSXGeomPlot( 'tmp/tmp.xml' );

RunOpenEMS('tmp','tmp.xml','');

Your support would by highly appreciated,
Thank you and regards,
ignitions2

**Comment by thliebig** [comment]
1) Well everything in openEMS is defined in basic SI units. Thus the electric fields are V/m and magnetic fields A/m and so on. I think you can add a label to the colorbar in Paraview, but not sure exactly where and how.
2) A lumped port is defined as two points yes, but these two points define a box. That box can be a line, a plane or a cube. Thus if it is a cube the direction of excitation (in other words the electric field (Ex vs Ey vs Ez) that gets excited). The amplitude is of course the strength of the excitation. The unit again is V/m. But I would have to check if the lumped port defines the excitation as a total voltage, then the electric field strength would depend on the port length...
3) I think there are plenty of examples/tutorials in openEMS to show this? But it basically works the same for every port...

I hope that helps?
br
Thorsten

**Reply by ignitions2** [reply]
Hi Thorsten,

Thank you for the prompt feedback.
Regarding the answers:

1. That is clear, thank you!
2. "But I would have to check if the lumped port defines the excitation as a total voltage, then the electric field strength would depend on the port length... " That would be exactly the question. Could you please check and confirm this point? Regarding this, it is also a question, whether these are "hard sources", i.e. the electric field is defined as a boundary condition, or "transparent sources", i.e. they don't interact with the scattered waves? Also, could you please link any literature, based on that I can understand how the ports are implemented?

Thank you in advance,
ignitions2

**Reply by thliebig** [reply]
Well this is all stuff you can check yourself very easily. The lumped port (like any port) records its port voltage. If that (in time domain) matched your excitation you know the excitation defines the port voltage.
But I'm pretty sure it directly defines the electric fields because it just simpler. In this case the port voltage is something like your excitation times the port length.
In any case, I usually never even care about all this, because in frequency domain you have to correctly normalize everything or you just look at relative values like S-parameters which is one voltage divided by another....

**Reply by thliebig** [reply]
I'm pretty sure there is a ton of literature on FDTD and such. But openEMS is not really implemented following any specific literature.
And I'm sure there are many ways to do this... If you really want to know the details, I recommend you study the source code, after all that is one of the reason why openEMS is OSS.

**Reply by LubomirJagos** [reply]
Port excitation is also question which I would be interested. So here is practical example what I tried with ignitions2 script and how to display port voltages, electric field strength:
    1. as Thorsten mentioned in FDTD excitation amplitude is amplitude of electric field which unit is V/m, therefore if your port is excited in Ex direction and length is 3.2mm to get 1V over it amplitude must be 1 / 3.2 * 1(V/mm) * 3.2(mm) = 1 / 3.2 * 1000(V/m) * 3.2(mm) so you have to define excitation amplitude in Ex as 1 / 3.2 * 1000 (V/m)
    2. to really have access to electric field you can use dumpbox in time domain and then display excitation amplitude in each recorded node as showed in script below
    
So I have looked at your script and correct it, there was lot of errors so here are main points what I did, I put comments into it what I made it to make it running properly:
    - operation frequency is 10GHz so lambda is 3cm, this is just for justification to have idea about dimentsions and what to expect
    - there are two parallel boxes like wires with 1x1mm approximately going in Z direction for 100mm so there should be at least like 3 peaks visible paraview for traverse wave
    - you used wrong material definition, wires now have using PEC
    - excitation amplitude change to 1/3.2*1000 V/m to get 1V at port, its impedance is 50Ohm, so current should be 20mA
    - simulation oversampling 50 removed
    - mesh made more coarse to get fast simulation
    - port must have HIGHER PRIORITY THAN striplines (higher priority means object is used when more objects overlapping)
    - port moved little into striplines, used offset 5mm then voltage is really like 1V at port
    - you originally defined unit as 1e-6 what are micrometers, don't think that that structure was meant to be in micrometers so I put right unit 1e-3 to have everything in milimeters what for 10GHz and wires long 10cm makes more sense to me

So I made simulation and at the end there is pos processing where from .h5 file electric field strength for port nodes is read and displayed in time, port contains 7x4x3 (X,Y,Z) nodes, so there are 3 graph in Z layer, you can clearly see at most top layer there is in middle electric field 1V in X direction and on lower layers there is significant also 0.5V in Z direction (so I think that's interaction with E field in time)

When reading E field for nodes I multiply it by 1e-3 = 1mm what is approximate distance of cell, I saw there is mesh information from h5 file, but don't know how to use it correctly to get right voltage, if somebody know can write here exact formula which distance I should take and multiply with E field intensity to get exact voltage, but 1mm is quite OK for this demonstration.

Here is corrected code:
```matlab
close all
clear all
clc

#######################################################################################
# INPUT PARAMETERS
#######################################################################################

output_dir = 'simulation_output_2';
output_filename = 'stripline.xml';
postprocessing_only = 0;

freq = [10e9]
stripline_length = 100;
port_Z_offset = 5;        #better results to reach 1V is when port is moved in Z direction


#######################################################################################
# SIMULATION SETTINGS
#######################################################################################

#unit = 1e-6;  #unit is um this is WRONG
unit = 1e-3;  #unit is mm

#FDTD = InitFDTD('NrTS',5e5, 'EndCriteria',0, 'OverSampling',50);
FDTD = InitFDTD('NrTS',10000, 'EndCriteria',0.01);

FDTD = SetSinusExcite(FDTD,freq);
#FDTD = SetBoundaryCond(FDTD,{'MUR' 'MUR' 'MUR' 'MUR' 'MUR' 'MUR'});
FDTD = SetBoundaryCond(FDTD,{'PML_8' 'PML_8' 'PML_8' 'PML_8' 'PML_8' 'PML_8'});

CSX = InitCSX();

#######################################################################################
# GRIDLINES
#######################################################################################

mesh.x = [            ...
  -10   :1.2   :-3    ...
  -2.8  :0.7   :2.8   ...
  3     :1.2   :10    ...
];
mesh.y = [
  -10   :1.2    :-2   ...
  -1.8  :0.7   :1.8   ...
  2     :1.2    :10   ...
];
mesh.z = -10: stripline_length+10;

CSX = DefineRectGrid(CSX, unit, mesh);

#######################################################################################
# MATERIAL
#######################################################################################

# WRONG, THIS IS SOME WEIRD MATERIAL
##CSX = AddMaterial( CSX, 'stripline1' );
##CSX = SetMaterialProperty(CSX, 'stripline1','Epsilon',1,'Mue',0.999991,'Kappa',59594756);

CSX = AddMetal( CSX, 'stripline1' );
start = [-2.6, -1, 0];
stop = [-0.6, 1, stripline_length];

# too high priority
#CSX = AddBox( CSX, 'stripline1', 999, start, stop );
CSX = AddBox( CSX, 'stripline1', 10, start, stop );

# WRONG, THIS IS SOME WEIRD MATERIAL
##CSX = AddMaterial( CSX, 'stripline2' );
##CSX = SetMaterialProperty(CSX, 'stripline2','Epsilon',1,'Mue',0.999991,'Kappa',59594756);

CSX = AddMetal( CSX, 'stripline2' );
start = [0.6, -1, 0];
stop = [2.6, 1, stripline_length];

# too high priority
#CSX = AddBox( CSX, 'stripline2', 999, start, stop );
CSX = AddBox( CSX, 'stripline2', 10, start, stop );




#######################################################################################
# PORT SETTINGS
#######################################################################################

#  WRONG, THIS IS PORT ON 1 GRIDLINE
#start = [-1.6 0 0];
#stop = [1.6 0 0];
#[CSX, port] = AddLumpedPort(CSX, 5 ,1 , 50, start, stop, [1e4 0 0], true);


#  RIGHT PUT PORT ON MORE GRIDLINES
start = [-1.6 -1 port_Z_offset+0];
stop = [1.6 1 port_Z_offset+2];

# PORT MUST HAVE HIGHER PRIORITY THAN STRIPLINES!!! MUST BE!
#[CSX, port] = AddLumpedPort(CSX, 20, 1 , 50, start, stop, [1e4 0 0], true);

#distacne is appr. 1mm so 1e3V/m*1mm = 1V
#[CSX, port] = AddLumpedPort(CSX, 20, 1 , inf, start, stop, [1e3 0 0], true);

# added type of excitation for port, this argument is defined in AddExcitation.m
#   - 'type': 1 = E hard excitation
#
# port is in X ax long 3.2mm so to got amplitude 3.2mm*1000V/m/3.2
[CSX, port] = AddLumpedPort(CSX, 20, 1 , 50, start, stop, [1/3.2*1e3 0 0], true, 'type', 0);



#######################################################################################
# DUMP FIELDS
#######################################################################################

%Checking E-field in x-z cross section
CSX = AddDump(CSX,'Et_x_z');
CSX = AddBox(CSX,'Et_x_z',0,[-10 0 -10],[10 0 stripline_length+10]);

%Checking E-field in x-y cross section
CSX = AddDump(CSX,'Et_x_y');
CSX = AddBox(CSX,'Et_x_y',0,[-10 -10 30],[10 10 30]);

%Checking B-field in x-y cross section
CSX = AddDump(CSX,'Bt_x_y','DumpType',5);
CSX = AddBox(CSX,'Bt_x_y',0,[-10 -10 30],[10 10 30]);

#
# ERROR: I think this is not implemnented and all data will be 0
#
%Checking current in x-y cross section
CSX = AddDump(CSX,'Jt_x_y','DumpType',0,'FileType', 1); #use hdf5 format
CSX = AddBox(CSX,'Jt_x_y',0,[-10 -10 30],[10 10 30]);


% add Et field probe for input port
start = [-1.6 -1 -1];
stop = [1.6 1 1];
CSX = AddDump(CSX,'port_ET','DumpType', 0, 'FileType', 1);  #use hdf5 format
CSX = AddBox(CSX,'port_ET',0,start,stop);



#######################################################################################
# SIMULATION RUN
#######################################################################################
try
  rmdir(output_dir, 's');
catch
  display(["Dir" output_dir "doesn't exists."]);
end_try_catch
mkdir(output_dir);

#
# Write simulation XML file.
#
WriteOpenEMS([output_dir, '/', output_filename],FDTD,CSX);
CSXGeomPlot( [output_dir, '/', output_filename]);

#
# Run openEMS if simulation enabled.
#
if postprocessing_only == 0
  RunOpenEMS(output_dir, output_filename,'');
end




#######################################################################################
# RESULT POST-PROCESSING
#######################################################################################
port = calcPort(port, output_dir, freq);

k = 1;
figure;
plot(port.ut.tot);
title("Voltage port in middle of dipole, Time Domain");
ylabel(['port ' num2str(k) ' voltage [V]']);
xlabel('time sample [-]');

figure;
plot(port.it.tot);
title("Current in middle of dipole, Time Domain");
ylabel(['port ' num2str(k) ' current [A]']);
xlabel('time sample [-]');

#dump read
[portEtField portEtMesh] = ReadHDF5Dump([output_dir '/port_ET.h5']);

#
# Choose node using indexing inside E field box, there is mesh variable which hold distancies between nodes
#
for zCoord = 1:3
  figure;
  title(["E-field at Z layer " num2str(zCoord)]);

  for row = 1:7
    for col = 1:3
      portEx = [];
      portEy = [];
      portEz = [];

      nodeIndexX = row;
      nodeIndexY = col;
      nodeIndexZ = zCoord;

      for k = 1:length(portEtField.TD.values)
        portEx = [portEx portEtField.TD.values{k}(nodeIndexX, nodeIndexY, nodeIndexZ,1)];
        portEy = [portEy portEtField.TD.values{k}(nodeIndexX, nodeIndexY, nodeIndexZ,2)];
        portEz = [portEz portEtField.TD.values{k}(nodeIndexX, nodeIndexY, nodeIndexZ,3)];
      end

      portEx = portEx .* 1e-3;
      portEy = portEy .* 1e-3;
      portEz = portEz .* 1e-3;

      subplot(7,3,3*(row-1)+col);
      hold on;
      ylim([-1, 1]);
      plot(portEx);
      plot(portEy);
      plot(portEz);

      legend("X","Y","Z")

  ##    figure;
  ##
  ##    portETotal = abs(sqrt(portEx.^2 + portEy.^2 + portEz.^2))
  ##    title("E field total value");
  ##    plot(portETotal)

    end
  end
end
```

Here is simulation and results:
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/8602e6d3-b673-40f0-989c-2630dbc2729e)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/1774232b-2b04-4811-b7ba-5ad26c119423)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/12271c12-5125-493a-9976-aacbfb945723)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/0af7c944-55bc-40dc-b585-a909a3109f16)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/4dd3582c-4574-426f-af4f-3c539c16f952)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/188af890-7f87-4390-8581-76333c795593)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/b2c09081-2b07-48b3-8c41-6b2963b57d17)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/4b909e46-c244-49e8-aa50-ecb8c5436e43)
![image](https://github.com/thliebig/openEMS-Project/assets/18229825/74c650f3-411e-4fbe-b586-5eb5000f55e0)

**Comment by ignitions2** [comment]
Hi LubomirJagos,

Thank you for your thorough answer regarding my question. Here are some comments below from my side.

"don't think that that structure was meant to be in micrometers so I put right unit 1e-3 to have everything in milimeters" --> the structure was intentionally defined in micrometers.
"you used wrong material definition, wires now have using PEC" --> actually PEC is what I would like to avoid for structure definition. The point would be to see skin depth, that is impossible with PEC, therefor I wanted to use structures having finite (but high) conductivity, therefor this was supposed to be two copper striplines. Is this the wrong way how I tried to define them?

This model is for microwave device application, therefore all dimensions and frequency were intentional.

Regards,
ignitions2
